From 30f60d2f71b3c14e09e82072f3da774c5bf5a94d Mon Sep 17 00:00:00 2001
From: Christophe Kerello <christophe.kerello@foss.st.com>
Date: Wed, 12 Jul 2023 18:15:55 +0200
Subject: [PATCH 055/120] spi: stm32_ospi: allow to sample later the data

Allow the data to be sampled later in order to consider
the external signal delays.
The calibration is done with and without TCR_SSHIFT and the best
configuration found is applied.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/u-boot.git 8a3ae40505d7f7932c4847ddab636d8e76c279a2 spi: stm32_ospi: allow to sample later the data]
Change-Id: I8b309739e60bc19d78a8e499d46509d628c1cd78
Signed-off-by: Christophe Kerello <christophe.kerello@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/u-boot/+/319450
Reviewed-by: Patrice CHOTARD <patrice.chotard@foss.st.com>
Tested-by: Christophe KERELLO <christophe.kerello@st.com>
Reviewed-by: Christophe KERELLO <christophe.kerello@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Domain-Review: Patrice CHOTARD <patrice.chotard@foss.st.com>
---
 drivers/memory/stm32-omi.c | 36 ++++++++++++++++++++++-
 drivers/spi/stm32_ospi.c   | 60 ++++++++++++++++++++++++++------------
 include/stm32_omi.h        |  4 ++-
 3 files changed, 80 insertions(+), 20 deletions(-)

diff --git a/drivers/memory/stm32-omi.c b/drivers/memory/stm32-omi.c
index 8c482a42183..17ab6df56a5 100644
--- a/drivers/memory/stm32-omi.c
+++ b/drivers/memory/stm32-omi.c
@@ -53,7 +53,7 @@ static int stm32_omi_dlyb_set_tap(struct udevice *dev, u8 tap, bool rx_tap)
 	return ret;
 }
 
-int stm32_omi_dlyb_find_tap(struct udevice *dev, bool rx_only)
+int stm32_omi_dlyb_find_tap(struct udevice *dev, bool rx_only, u8 *window_len)
 {
 	struct stm32_omi_priv *omi_priv = dev_get_priv(dev);
 	struct stm32_tap_window rx_tap_w[DLYBOS_TAPSEL_NB];
@@ -118,6 +118,7 @@ int stm32_omi_dlyb_find_tap(struct udevice *dev, bool rx_only)
 		}
 
 		rx_tap = rx_window_end - rx_window_len / 2;
+		*window_len = rx_window_len;
 		dev_dbg(dev, "RX_TAP_SEL set to %d\n", rx_tap);
 
 		return stm32_omi_dlyb_set_tap(dev, rx_tap, true);
@@ -153,6 +154,39 @@ int stm32_omi_dlyb_find_tap(struct udevice *dev, bool rx_only)
 	return stm32_omi_dlyb_set_tap(dev, best_tx_tap, false);
 }
 
+int stm32_omi_dlyb_set_cr(struct udevice *dev, u32 dlyb_cr)
+{
+	bool bypass_mode = false;
+	int ret;
+	u16 period_ps;
+	u8 rx_tap, tx_tap;
+
+	period_ps = FIELD_GET(DLYBOS_BYP_CMD_MASK, dlyb_cr);
+	if (dlyb_cr & DLYBOS_BYP_EN)
+		bypass_mode = true;
+
+	ret = stm32_omi_dlyb_configure(dev, bypass_mode, period_ps);
+	if (ret)
+		return ret;
+
+	/* restore Rx and TX tap */
+	rx_tap = FIELD_GET(DLYBOS_CR_RXTAPSEL_MASK, dlyb_cr);
+	ret = stm32_omi_dlyb_set_tap(dev, rx_tap, true);
+	if (ret)
+		return ret;
+
+	tx_tap = FIELD_GET(DLYBOS_CR_TXTAPSEL_MASK, dlyb_cr);
+	return stm32_omi_dlyb_set_tap(dev, tx_tap, false);
+}
+
+void stm32_omi_dlyb_get_cr(struct udevice *dev, u32 *dlyb_cr)
+{
+	struct stm32_omi_plat *omi_plat = dev_get_plat(dev);
+
+	regmap_read(omi_plat->regmap, omi_plat->dlyb_base + SYSCFG_DLYBOS_CR,
+		    dlyb_cr);
+}
+
 /* ½ memory clock period in pico second */
 static const u16 dlybos_delay_ps[STM32_DLYBOS_DELAY_NB] = {
 2816, 4672, 6272, 7872, 9472, 11104, 12704, 14304, 15904, 17536, 19136, 20736,
diff --git a/drivers/spi/stm32_ospi.c b/drivers/spi/stm32_ospi.c
index 49229c82976..8ee1b3d1e70 100644
--- a/drivers/spi/stm32_ospi.c
+++ b/drivers/spi/stm32_ospi.c
@@ -21,7 +21,6 @@ struct stm32_ospi_priv {
 	struct udevice *omi_dev;
 	int cs_used;
 	u64 idcode;
-	u32 prescaler[OSPI_MAX_CHIP];
 };
 
 static int stm32_ospi_mm(struct udevice *omi_dev,
@@ -231,9 +230,6 @@ static int stm32_ospi_probe(struct udevice *bus)
 	reset_deassert_bulk(&omi_plat->rst_ctl);
 
 	priv->cs_used = -1;
-	memset(priv->prescaler, -1, OSPI_MAX_CHIP * sizeof(u32));
-
-	setbits_le32(regs_base + OSPI_TCR, OSPI_TCR_SSHIFT);
 
 	/* Set dcr devsize to max address */
 	setbits_le32(regs_base + OSPI_DCR1, OSPI_DCR1_DEVSIZE_MASK);
@@ -288,7 +284,11 @@ static int stm32_ospi_set_speed(struct udevice *bus, uint speed)
 static int stm32_ospi_calibration(struct udevice *bus, uint freq)
 {
 	struct stm32_ospi_priv *priv = dev_get_priv(bus);
-	int ret;
+	struct stm32_omi_plat *omi_plat = dev_get_plat(priv->omi_dev);
+	phys_addr_t regs_base = omi_plat->regs_base;
+	u32 dlyb_cr;
+	u8 window_len_tcr0 = 0, window_len_tcr1 = 0;
+	int ret, ret_tcr0, ret_tcr1;
 
 	/*
 	 * Set memory device at low frequency (50MHz) and sent
@@ -312,12 +312,39 @@ static int stm32_ospi_calibration(struct udevice *bus, uint freq)
 	if (ret)
 		return ret;
 
-	/* perform only RX TAP selection */
-	ret = stm32_omi_dlyb_find_tap(priv->omi_dev, true);
+	ret_tcr0 = stm32_omi_dlyb_find_tap(priv->omi_dev, true, &window_len_tcr0);
+	if (!ret_tcr0)
+		stm32_omi_dlyb_get_cr(priv->omi_dev, &dlyb_cr);
+
+	ret = stm32_omi_dlyb_stop(priv->omi_dev);
 	if (ret)
+		return ret;
+
+	ret = stm32_omi_dlyb_configure(priv->omi_dev, false, 0);
+	if (ret)
+		return ret;
+
+	setbits_le32(regs_base + OSPI_TCR, OSPI_TCR_SSHIFT);
+
+	ret_tcr1 = stm32_omi_dlyb_find_tap(priv->omi_dev, true, &window_len_tcr1);
+	if (ret_tcr0 && ret_tcr1) {
 		dev_info(bus, "Calibration phase failed\n");
+		return ret_tcr0;
+	}
 
-	return ret;
+	if (window_len_tcr0 >= window_len_tcr1) {
+		clrbits_le32(regs_base + OSPI_TCR, OSPI_TCR_SSHIFT);
+
+		ret = stm32_omi_dlyb_stop(priv->omi_dev);
+		if (ret)
+			return ret;
+
+		ret = stm32_omi_dlyb_set_cr(priv->omi_dev, dlyb_cr);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
 }
 
 static int stm32_ospi_claim_bus(struct udevice *dev)
@@ -330,7 +357,7 @@ static int stm32_ospi_claim_bus(struct udevice *dev)
 	u32 dcr2, prescaler;
 	int slave_cs = slave_plat->cs[0];
 	uint bus_freq;
-	int ret = 0;
+	int ret;
 
 	if (slave_cs >= OSPI_MAX_CHIP)
 		return -ENODEV;
@@ -342,21 +369,17 @@ static int stm32_ospi_claim_bus(struct udevice *dev)
 
 	priv->cs_used = slave_cs;
 
+	ret = stm32_omi_dlyb_stop(priv->omi_dev);
+	if (ret)
+		return ret;
+
 	/* Set chip select */
 	clrsetbits_le32(regs_base + OSPI_CR, OSPI_CR_CSSEL,
 			priv->cs_used ? OSPI_CR_CSSEL : 0);
+	clrbits_le32(regs_base + OSPI_TCR, OSPI_TCR_SSHIFT);
 
 	dcr2 = readl(regs_base + OSPI_DCR2);
 	prescaler = (dcr2 & OSPI_DCR2_PRESC_MASK) >> OSPI_DCR2_PRESC_SHIFT;
-
-	if (prescaler == priv->prescaler[priv->cs_used])
-		return 0;
-
-	ret = stm32_omi_dlyb_stop(priv->omi_dev);
-	if (ret)
-		return ret;
-
-	priv->prescaler[priv->cs_used] = prescaler;
 	bus_freq = ospi_clk / (prescaler + 1);
 
 	/* calibration needed above 50MHz */
@@ -370,6 +393,7 @@ static int stm32_ospi_claim_bus(struct udevice *dev)
 			if (ret)
 				return ret;
 
+			clrbits_le32(regs_base + OSPI_TCR, OSPI_TCR_SSHIFT);
 			ret = stm32_ospi_set_speed(dev->parent, STM32_DLYB_FREQ_THRESHOLD);
 		}
 	}
diff --git a/include/stm32_omi.h b/include/stm32_omi.h
index e64cb15e991..b35d21134b3 100644
--- a/include/stm32_omi.h
+++ b/include/stm32_omi.h
@@ -166,7 +166,9 @@ struct stm32_tap_window {
 
 int stm32_omi_dlyb_configure(struct udevice *dev,
 			     bool bypass_mode, u16 period_ps);
-int stm32_omi_dlyb_find_tap(struct udevice *dev, bool rx_only);
+int stm32_omi_dlyb_find_tap(struct udevice *dev, bool rx_only, u8 *window_len);
+int stm32_omi_dlyb_set_cr(struct udevice *dev, u32 dlyb_cr);
+void stm32_omi_dlyb_get_cr(struct udevice *dev, u32 *dlyb_cr);
 int stm32_omi_dlyb_stop(struct udevice *dev);
 int stm32_omi_tx_poll(struct udevice *dev, u8 *buf, u32 len, bool read);
 int stm32_omi_wait_cmd(struct udevice *dev);
-- 
2.47.2

